proposals/006-spinapp-invocation-limits/README.md

```diff
@@ -0,0 +1,66 @@
+# SKIP 006 - SpinApp invocation limits
+
+Summary: Add support for configuring per-invocation resource limits in `SpinApp` CRD
+
+Owner: Kate Goldenring <kate.goldenring@fermyon.com>
+
+Impacted Projects:
+
+- [x] spin-operator
+- [ ] `spin kube` plugin
+- [ ] runtime-class-manager
+- [ ] containerd-shim-spin
+- [ ] Governance
+- [ ] Creates a new project
+
+Created: 05-29-2025
+
+Updated: 05-29-2025
+
+## Background
+
+Spin recently [added support for limiting instance memory](https://github.com/spinframework/spin/pull/3135) and has had an outstanding request for supporting [request timeouts](https://github.com/spinframework/spin/issues/30). The `SpinApp` CRD should support configuring per invocation limits.
+
+## Proposal
+
+Add an `invocationLimits` section to the `SpinApp` CRD with a mapping of limits. Executors can define valid limits that they support.
+
+For example, if the `foo` executor supports  `maxMemory` and `maxExecutionTime` limits, the following `SpinApp` custom resource could be applied:
+
+```yaml
+apiVersion: core.spinkube.dev/v1alpha1
+kind: SpinApp
+metadata:
+  name: hello
+spec:
+  executor: foo
+  image: "ghcr.io/spinframework/myapp:latest"
+  replicas: 1
+  invocationLimits:
+    maxMemory: "64Mi"
+    maxExecutionTime: "30s"
+```
+
+For the Spin containerd shim executor, the Spin Operator can configure the per invocation memory limit by setting the (converted to bytes) value of the `maxMemory` in the `SPIN_MAX_INSTANCE_MEMORY` environment variable on Deployments. This is similar to how OTEL settings [are configured](https://github.com/spinframework/spin-operator/blob/4900cf4bdfdc519c152e7d20f2529fccee7ba40f/internal/controller/deployment.go#L128would).
+
+```yaml
+apiVersion: core.spinkube.dev/v1alpha1
+kind: SpinApp
+metadata:
+  name: hello
+spec:
+  executor: containerd-shim-spin
+  image: "ghcr.io/spinframework/myapp:latest"
+  replicas: 1
+  # Pod level resource limits
+  resources:
+    limits:
+      memory: "128Mi"
```

Comment:
does there need to be any correlation to make sure the limit of the pod is greater than the invocation limit?  What happens if resource limit is lower than the invocation limit?

Reply:
That would be a good check. We could have the operator validate that for the shim executor. Not all executors are pod based and support pod limits

Reply:
> Not all executors are pod based and support pod limits

good to know.  What are other executors? Could they have limits on their environment as well?

> This feels like an implementation detail, but do you think it is important to point out the relationship between pod and invocation limits and this validation that would happen in the operator?

It seems like an user friction spot. Maybe there is another webhook that sets pod limits for pods and the user doesn't know this and gets confused that their invocation limits never get set? or maybe they never find out.  Elevating a warning at the least seems reasonable?

In the case of the pod executor, Is it a bad thing if the user does this? could it just be a condition on the pod as a warning?